content/00.zksync-network/50.sdk/10.js/01.ethers/04.guides/04.interop-messages.md

```diff
@@ -0,0 +1,121 @@
+
+The message itself has no destination chain or address.
+It’s simply a payload created by a user or contract that gets broadcast.
+There is no expiration for when messages can be verified.
+
+Messages can only be sent and verified on ZKsync chains that use ZKsync Gateway.
+To set up a local testing environment with Gateway, check out the [Running ZKsync Gateway locally guide](/zk-stack/running/gateway-settlement-layer).
+
+Before verifying a message, the transaction that the message was sent in must be executed in a batch,
+and the interop root must be updated on the target chain.
+The full flow for sending and verifying a message looks like this:
+
+1. Send the message.
+1. Check that the message transaction is fully executed.
+1. Check that the gateway proof is ready.
+1. Check that the interop root is updated on the target chain.
+1. Verify the message.
+
+## Sending a message
+
+You can use the `InteropClient` along with the `sendMessage` method to send any message.
+The method accepts either a string or a `BytesLike` value, and returns the transaction hash.
+
+```ts
+:code-import{filePath="hardhat-sol/scripts/interop/send-message.ts"}
+```
+
+### Sending a message from a smart contract
+
+To send a message from a smart contract, you can use the `sendToL1` method on the `L1Messenger` contract.
+
+```solidity
+:code-import{filePath="hardhat-sol/contracts/InteropSendMessage.sol"}
+```
+
+## Checking the execution status
+
+To check the execution status of a message, you can use the `getMessageStatus` method.
+In order to verify a message, the status must first be `EXECUTED`.
+
+```ts
+:code-import{filePath="hardhat-sol/scripts/interop/check-status.ts"}
+```
+
+## Checking if the interop root is updated
+
+In order to verify a message, the interop root must be updated on the target chain.
+To check if the interop root is ready, you can use the `getGwBlockForBatch` `waitForGatewayInteropRoot` methods.
+
+```ts
+:code-import{filePath="hardhat-sol/scripts/interop/check-interop-root.ts"}
+```
+
+## Verifying a Message
+
+You can either verify a message using the SDK or onchain inside a smart contract.
+
+### Offchain Verification
+
+You can verify a message directly using the SDK by calling the `verifyMessage` method on the `InteropClient`.
+
+```ts
+:code-import{filePath="hardhat-sol/scripts/interop/verify-message.ts"}
+```
+
+To verify a message on a local chain,
+you must also send some transactions on the target chain in order to create a new batch so that the interop root gets updated.
+You can see an example of this below under "Local chains example".
+
+### Onchain Verification
+
+To verify a message inside a smart contract,
+you can fetch the input args for `proveL2MessageInclusionShared` using the `getVerificationArgs` method.
+
+### Testnet example
+
+```ts
+:code-import{filePath="hardhat-sol/scripts/interop/get-verification-args.ts"}
+```
+
+### Local chains example
+
+To verify a message on a local chain,
```

Comment:
Do we have anywhere we can link out for instruction on setting up a multichain local env?

Reply:
not yet, but I can add